pkg/clientcert/cert_controller.go

```diff
@@ -0,0 +1,317 @@
+
+	ClusterNameLabel = "open-cluster-management.io/cluster-name"
+	AddonNameLabel   = "open-cluster-management.io/addon-name"
+	SignerNameLabel  = "open-cluster-management.io/signer-name"
+)
+
+// ControllerResyncInterval is exposed so that integration tests can crank up the constroller sync speed.
+var ControllerResyncInterval = 5 * time.Minute
+
+// NewCSRFunc returns a csr with custom name (or generated name), labels and signer.
+type NewCSRFunc func(subject *pkix.Name, keyData []byte) (*certificates.CertificateSigningRequest, error)
+
+// ExtendClientCertSecretFunc extends the data of client certificate secret created by controller
+// with addional data. The input data may include keys: tls.key and tls.crt. A new map should be returned
+// and it should include the original data and perhaps the new keys, like kubeconfig.
+type ExtendClientCertSecretDataFunc func(data map[string][]byte) (map[string][]byte, error)
+
+// clientCertificateController implements the common logic of hub client certification creation/rotation. It
+// creates a client certificate and rotates it before it becomes expired by using csrs. The client
+// certificate generated is stored in a specific secret with the keys below:
+// 1). tls.key: tls key file
+// 2). tls.crt: tls cert file
+type clientCertificateController struct {
+	clientCertSecretNamespace      string
+	clientCertSecretName           string
+	hubCSRLister                   certificateslisters.CertificateSigningRequestLister
+	hubCSRClient                   csrclient.CertificateSigningRequestInterface
+	spokeCoreClient                corev1client.CoreV1Interface
+	csrSubject                     *pkix.Name
+	newCSRFunc                     NewCSRFunc
+	extendClientCertSecretDataFunc ExtendClientCertSecretDataFunc
+	controllerName                 string
+
+	// csrName is the name of csr created by controller and waiting for approval.
+	csrName string
+
+	// keyData is the private key data used to created a csr
+	// csrName and keyData store the internal state of the controller. They are set after controller creates a new csr
+	// and cleared once the csr is approved and processed by controller. There are 4 combination of their values:
+	//   1. csrName empty, keyData empty: means we aren't trying to create a new client cert, our current one is valid
+	//   2. csrName set, keyData empty: there was bug
+	//   3. csrName set, keyData set: we are waiting for a new cert to be signed.
+	//   4. csrName empty, keydata set: the CSR failed to create, this shouldn't happen, it's a bug.
+	keyData []byte
+}
+
+func NewClientCertificateController(
+	clientCertSecretNamespace string,
+	clientCertSecretName string,
+	hubCSRInformer certificatesinformers.CertificateSigningRequestInformer,
+	hubCSRClient csrclient.CertificateSigningRequestInterface,
+	spokeSecretInformer corev1informers.SecretInformer,
+	spokeCoreClient corev1client.CoreV1Interface,
+	csrSubject *pkix.Name,
+	newCSRFunc NewCSRFunc,
```

Comment:
If the only reason that we have this func is because of signer, i would directly use signer as param here.

Reply:
When creating a csr, the following parts can be customized:
- name/generated name;
- labels
- subject (cn, o, ou...)
- signer

Besides subject, we still need 3 parameters to cover all of them. So passing a function might be better in this situation.

Reply:
I replaced the func with 3 parameters below:
```
csrObjectMeta metav1.ObjectMeta
csrSubject    *pkix.Name
csrSignerName string
```